### DISCUSSION TITLE
AWS: Make Fetches Concurrent

### DISCUSSION THREAD

**Discussion by ramonpetgrave64** [cartography-cncf/cartography#1201] [Ideas]
*Title*: *Make Fetches Concurrent*

*Description*:
Cartography up to now is entirely serial. But there are many opportunities for concurrency, particularly during the `get_` phases of sync modules.

As a basic example: To get all the data needed for all AWS ECR Images, we:
1. call boto's method, to get a list of all repositories
1. for each repository, we call another boto method to list all images within that repository

The loop in the second step above is what we can make concurrent, saving time of network overhead when there are perhaps many thousands of repositories.

 The[ boto clients are also thread-safe](https://boto3.amazonaws.com/v1/documentation/api/1.19.0/guide/clients.html#multithreading-or-multiprocessing-with-clients), meaning it's okay to run them in multiple concurrent threads.

Ideally, we would use the async version of client libraries, but boto does not support that. There is a project [aiboto3](https://pypi.org/project/aioboto3/) that seems to wrap around boto to make its methods async, but personally I would rather not.

I have a pending PR as another workaround #1192. The idea there is to do minimal refactoring by 
1. converting our loops to async methods
2. calling boto's get methods from separate threads

To be more concrete with the ECR Images example: you might think to just convert `get_ecr_repository_images` to an `async def`, but that still does not yield any performance gain because inside of that is a long-running function call with boto's `list_images` method. Because of the GIL, it is ultimately the `list_images` method that needs to run in a separate thread, but I chose to keep the refactor simple by putting it's only caller in a separate thread.

I've read a bit about asyncio and GIL from the realpython guides, but I'm still new and would like to hear thoughts from the community.

*Relevant Links*:

* #1192 
* https://realpython.com/python-gil/
* https://realpython.com/python-async-features/
* https://realpython.com/async-io-python/
* https://boto3.amazonaws.com/v1/documentation/api/1.19.0/guide/clients.html#multithreading-or-multiprocessing-with-clients

**Comment by achantavy** [comment]
Some things to consider:

How will we apply/refactor this change to all AWS modules?

Will it be easy for new module authors to copy this pattern in their own modules?

Can you describe the performance gain that you observed in #1192 with a bit more detail?

**Reply by ramonpetgrave64** [reply]
2. Not difficult. Mainly just converting loop bodies. The hardest part will be adding tests where they didn't exist before. But I think manual testing will be enough.
3. The s3 and ec3 syncs would take about 5 and 15 minutes, respectively. With the async, they now take about 35 seconds and 2 minutes, respectively. 
1. It could also be applied to the other non-aws modules. But for aws, it will be a similar process, as in the PR. Here are the candidates I've identified, but we can open a new GH issue/project to track the effort:

    - [dynamodb](https://github.com/lyft/cartography/blob/143a443fd1a62bff8abb561cb365376a91c16b22/cartography/intel/aws/dynamodb.py#L30-L31)
    - [ecs](https://github.com/lyft/cartography/blob/143a443fd1a62bff8abb561cb365376a91c16b22/cartography/intel/aws/ecs.py#L43)
    - [eks](https://github.com/lyft/cartography/blob/143a443fd1a62bff8abb561cb365376a91c16b22/cartography/intel/aws/eks.py#L109-L110)
    - [elasticsearch](https://github.com/lyft/cartography/blob/143a443fd1a62bff8abb561cb365376a91c16b22/cartography/intel/aws/elasticsearch.py#L42-L43)
    - [emr](https://github.com/lyft/cartography/blob/143a443fd1a62bff8abb561cb365376a91c16b22/cartography/intel/aws/emr.py#L91)
    - [iam](https://github.com/lyft/cartography/blob/143a443fd1a62bff8abb561cb365376a91c16b22/cartography/intel/aws/iam.py)
      - This will be a difficult refactor
    - [kms](https://github.com/lyft/cartography/blob/143a443fd1a62bff8abb561cb365376a91c16b22/cartography/intel/aws/kms.py#L47)
    - [lambda](https://github.com/lyft/cartography/blob/143a443fd1a62bff8abb561cb365376a91c16b22/cartography/intel/aws/lambda_function.py#L111)
    - [resourcegroupstaggingapi](https://github.com/lyft/cartography/blob/143a443fd1a62bff8abb561cb365376a91c16b22/cartography/intel/aws/resourcegroupstaggingapi.py#L203-L213)
    - [route53](https://github.com/lyft/cartography/blob/143a443fd1a62bff8abb561cb365376a91c16b22/cartography/intel/aws/route53.py#L395-L396)
    - [sqs](https://github.com/lyft/cartography/blob/143a443fd1a62bff8abb561cb365376a91c16b22/cartography/intel/aws/sqs.py#L42)
    - [ssm](https://github.com/lyft/cartography/blob/143a443fd1a62bff8abb561cb365376a91c16b22/cartography/intel/aws/ssm.py#L39)

**Comment by mpurusottamc** [comment]
We tried to incorporate multi-threading. One of the challenges we faced was around Rate Limits. boto3 has retry policies in place. But, it sometimes fails if there are a ton of concurrent requests going to AWS.

**Reply by ramonpetgrave64** [reply]
Yes, I've seen Throttling exceptions, but I've mitigated with the `@backoff` decorator with its `expo` waiter. Ratelimits vary by AWS service and each individual API, and they don't always publish those limits.

For example, I recently did some one-off async calls against iam's `generate_service_last_accessed_details` and `get_service_last_accessed_details`, where I did get throttling exceptons, but the `@backoff` decorator against that exception was enough, and I still got results in ~20% of the usual time without async.

Also, the default threadpool I'm proposing would only allow a maximum of 10 concurrent requests, though we could increase that for more gains.

**Reply by ramonpetgrave64** [reply]
Can you tell us about your own approach at multithreading?

**Reply by ramonpetgrave64** [reply]
@mpurusottamc I've just updated the PR to use @backoff builtin. Please take a look.
